Replace `libc` with `cty`

denzp opened the pull request:
The main reason for the PR is that **libc** is not `no_std` with default features, so it was impossible to use math.rs in on architectures that don't have `std`.

Since we need only C types from **libc**, it's reasonable to replace it with more lightweight lib.

nagisa commented:
This seems fairly okay to me, although I’m interested in why did you bind specifically to `0.1.4` as opposed to a more relaxed but equally valid bound of `0.1`?

denzp commented:
Ah, you right. Thanks for the point! I had to bump version couple times and was just glad it works ;)

I'll update the PR for more relaxed bound.

nagisa commented on src/utils.rs:
```diff
@@ -85,4 +135,4 @@ pub const FRAC_3PI_4: f64 = 2.35619449019234492884698253745962716314787704953132
 pub const FRAC_3PI_2: f64 = 4.7123889803846898576939650749192543262957540990626587;
 pub const FRAC_7PI_4: f64 = 5.4977871437821381673096259207391300473450464489064351;
-pub const PI_2: f64 = 6.2831853071795864769252867665590057683943387987502116;
+pub const PI_2: f64 = 6.2831853071795864769252867665590057683943387987502116;
```
This line introduces a whitespace error.

nagisa reviewed (commented):
There is a fair amount of changes unrelated to using cty in this PR (the abs changes for example).

Please either split these out into a separate PR or provide the motivation for these in the commit messages.

denzp commented:
I didn't intend to make any changes except replacing `libc` with `cty`, but I had to... Without `libc`, which is not `no_std` crate, we don't have needed methods, like `f64::abs` or `f64::sqrt`:
```
error: no method named `abs` found for type `f64` in the current scope
  --> src/acos.rs:11:28
   |
11 |         FRAC_PI_2 + asin(i.abs())
   |                            ^^^
   |
   = help: items from traits can only be used if the trait is in scope; the following trait is implemented but not in scope, perhaps add a `use` for it:
   = help: candidate #1: `use core::num::Float;`

error: no method named `sqrt` found for type `f64` in the current scope
  --> src/hypot.rs:35:25
   |
35 |     (dl * dl + dr * dr).sqrt() as f32
   |                         ^^^^
```

I'll rework the PR into 2 commits:
1. Move from `libc` to `cty` without `#![no_std]`
2. Enable `#![no_std]` again

nagisa commented:
You could’ve simply added the methods in question to the Float trait, the same thing you did for `is_nan` et al.

denzp committed:
Use `cty` instead of `libc`

This also requires us to implement own `utils::Float` trait for {f64, f32} due the lack of `std` create.
But since "test" target uses thirdparty non `no_std` crates and has `std`, our `utils::Float` is useless for the target.

denzp commented:
Thanks, it is a great idea in general. The only problem I see: **test** target has some thirdparty crates which provide `std` there. So to avoid warnings, I had to use:
```
#[cfg(not(test))]
use utils::Float;
```

But it also means, if `utils::Float` methods will have different behaviour from `std::{f32, f64}::*` methods, tested code might not reflect real code.

nagisa commented:
Thank you! I replaced the mutiple `cfg`s with a single that `cfg`s out the whole crate if it is compiled with `--test`, because it is cleaner that way and we do not have any tests in the library itself, only in the separate files.

denzp commented:
Ok, thanks a lot!
